endpoints.tf

```
@@ -13,3 +13,147 @@ resource "aws_vpc_endpoint" "transfer_server" {
   vpc_endpoint_type   = "Interface"
   vpc_id              = aws_vpc.default.id
 }
+
+
+data "aws_region" "current" {}
+
+resource "aws_vpc_endpoint" "s3" {
+  count             = var.private_s3_endpoint != false ? 1 : 0
+  service_name      = "com.amazonaws.${data.aws_region.current.name}.s3"
+  tags              = merge(var.tags, { "Name" = "s3-vpc-endpoint-${var.name}" })
+  vpc_endpoint_type = "Gateway"
+  vpc_id            = aws_vpc.default.id
+  route_table_ids   = aws_route_table.private[*].id
+}
+
+data "aws_vpc_endpoint_service" "ssm_endpoint" {
+  count   = length(var.ssm_endpoint.subnet_ids) != 0 ? 1 : 0
```

As per the comment already made by @shoekstra, prefer `null` to indicate something should not be set/used. So this would then become:
```suggestion
  count   = var.ssm_endpoint != null ? 1 : 0
```

---

Ok, so I have the count on length because to use an ssm_endpoint you MUST give subnet ids (as its an interface, no interface in a subnet, no endpoint). Therefore I can do this 
```

variable "ssm_endpoint" {
  type = object({
    private_dns_enabled = bool
    security_group_ids  = list(string)
    subnet_ids          = list(string)
  })
  default = {
    private_dns_enabled = false,
    security_group_ids  = [],
    subnet_ids          = []
  }
  description = "Variables to provision an ssm endpoint to the VPC"
}
```

so the defaults are actually set.

The reason I'm setting defaults is that I want to enable two things 

1. choosing to deploy an endpoint or not (and those defaults will result in 'not')
2. but I also want to allow the ability to pass already created security groups to the module (endpoint creation can take more than one if you want). However, it's easier usually to not pass any security groups at which point the module will create them. 

Doing it like this is the only way I could make terraform happy, because I was trying to use 'null' and then 

`output "ssmmessages_endpoint_security_group_ids" {
  value = length(var.ssmmessages_endpoint.security_group_ids) == 0 ? aws_security_group.ssmmessages_endpoint.*.id : var.ssmmessages_endpoint.security_group_ids
}
`
the outputs would complain as they types based on the conditional output were not equivalent.

tl;dr 
I'm defaulting the values to not null, because I want conditional behaviour for the security groups and its the only way I could get it to work. 

Any pointers on how to fix that another way would be welcome as I fought with it for quite a while

---

Inputs to resources are also exported, so you could first test if your variable is set and either print an empty string or the value passed to the resource, something like this::

```hcl
output "ssmmessages_endpoint_security_group_ids" {
  value = var.ssmmessages_endpoint == null ? "" : aws_vpc_endpoint.ssmmessages_endpoint.*.security_group_ids
}
```

Let's discuss to work out best way forward.